controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +70,211 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
 func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	policy := &policyv1beta1.OperatorPolicy{}
 
-	err := r.Get(ctx, req.NamespacedName, policy)
+	watcher := depclient.ObjectIdentifier{
+		Group:     operatorv1.GroupVersion.Group,
+		Version:   operatorv1.GroupVersion.Version,
+		Kind:      "OperatorPolicy",
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}
+
+	// Start query batch for caching and watching related objects
+	err := r.DynamicWatcher.StartQueryBatch(watcher)
+	if err != nil {
+		panic(err)
+	}
+
+	// Get the applied OperatorPolicy
+	err = r.Get(ctx, req.NamespacedName, policy)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			OpLog.Info("Operator policy could not be found")
 
+			err := r.DynamicWatcher.EndQueryBatch(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error ending the query batch in progress. Ignoring the failure.")
+			}
+
+			err = r.DynamicWatcher.RemoveWatcher(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error updating dependency watcher. Ignoring the failure.")
+			}
+
 			return reconcile.Result{}, nil
 		}
+
+		OpLog.Error(err, "Failed to get operator policy")
+
+		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, nil
-}
+	// handle the policy
+	OpLog.Info("Reconciling OperatorPolicy", "policy", policy.Name)
 
-// PeriodicallyExecOperatorPolicies loops through all operatorpolicies in the target namespace and triggers
-// template handling for each one. This function drives all the work the operator policy controller does.
-func (r *OperatorPolicyReconciler) PeriodicallyExecOperatorPolicies(freq uint, elected <-chan struct{},
-) {
-	OpLog.Info("Waiting for leader election before periodically evaluating operator policies")
-	<-elected
+	// Check if specified Subscription exists
+	cachedItem, err := r.DynamicWatcher.Get(watcher, subscriptiongvk, policy.Spec.Subscription.Namespace,
```

Comment:
From what I can tell, `policy.Spec.Subscription` is not a required field:
https://github.com/open-cluster-management-io/config-policy-controller/blob/7549739f09cafec987a292aac32925718f8b1561/api/v1beta1/operatorpolicy_types.go#L107

Should it be? If not, then this code needs to guard against that not being set.

Reply:
That's an good point. I think you mentioned the capability of the controller to be able to handle operator installations through means other than the OperatorPolicy. In that case, and in the general case, I think the subscription field must be set to ensure the subscription is either being created, and/or watched. I'll make this change.